### PR TITLE
Improved transparent rendering

### PR DESCRIPTION
**Pull request by elalish**
This is a follow-on to #22425 and #22473 to make rendering transparent objects onto a transparent canvas work better so they can be composited onto DOM or AR content behind and look reasonably seamless. This is required as DOM-to-texture is not and will not be allowed due to security, and WebXR does not expose the camera feed by default, meaning those backgrounds are not available within our shaders. It also means what I'm doing is fundamentally a non-PBR hack, since blend functions can't represent light transport properly, and it's not possible to handle transmission roughness or refraction. However, we can improve significantly:

Before / Current (including #25881)
imageimage
(for historical reference) 1st draft / 2nd draft:
imageimage

Before / After:
imageimage
And it works with other CSS background colors since the canvas is transparent:
imageimageimageimage

~~It does require two extra steps to make it work in addition to this PR:~~
~~1) The renderer must have `premultipliedAlpha: false`~~
~~2) Must call `renderer.setClearColor(new Color(0xFFFFFF), 0);`~~

~~I haven't included these in the spirit of avoiding breaking changes and because this unblocks model-viewer. However, it might be worth looking at these defaults, as materials default `premultipliedAlpha` to `false`, while the renderer defaults to `true`. Also, if you set `renderer.setClearColor(new Color(0xFFFFFF));`, it will ignore the `alpha` of your context and clear to `alpha = 1`, which is a bit confusing. Currently the background clears to black by default, but white seems just as reasonable?~~

### CODE REVIEW

File: src/renderers/shaders/ShaderChunk/transmission_pars_fragment.glsl.js [hunk 1]
```diff
@@ -184,12 +184,15 @@ export default /* glsl */`
...
+		vec3 transmittance = diffuseColor * volumeAttenuation( length( transmissionRay ), attenuationColor, attenuationDistance );
+		vec3 attenuatedColor = transmittedLight.rgb * transmittance;
 
 		// Get the specular component.
 		vec3 F = EnvironmentBRDF( n, v, specularColor, specularF90, roughness );
 
-		return vec4( ( 1.0 - F ) * attenuatedColor * diffuseColor, transmittedLight.a );
+		float transmittanceFactor = ( transmittance.r + transmittance.g + transmittance.b ) / 3.0;
+
+		return vec4( ( 1.0 - F ) * attenuatedColor, 1.0 - ( 1.0 - transmittedLight.a ) * transmittanceFactor );
```
Comment (elalish) [inline src/renderers/shaders/ShaderChunk/transmission_pars_fragment.glsl.js:195]: Here's the key to my new hack, which I feel is more defensible than what it replaces. As before, when `transmittedLight.a == 1` (the scene is opaque), no change to the proper PBR computation occurs. As this transmissive material gets lighter and the scene behind gets more transparent, output `alpha` decreases to blend in more of the unknown CSS background color.

**Comment by WestLangley** [comment]
>The renderer must have `premultipliedAlpha: false`

In theory, it can't, without breaking normal blending.

The three.js normal blending function outputs premultiplied values, hence `renderer.premultipliedAlpha` must remain `true` so that the drawing buffer is composited with the canvas background correctly. 

>However, it might be worth looking at these defaults, as _materials_ default premultipliedAlpha to false, while the _renderer_ defaults to true.

They are completely different properties, with different purposes. There is no reason why they should be the same.

**Comment by elalish** [comment]
@WestLangley Interesting, thanks! I just read up on premultiplied alpha again - makes me think I should try and tweak this approach so that I can keep that `true`. What blocked me is that transmission samples from the drawing buffer to determine incoming light and transparency. On the first pass, all that's there is the clear color, which we need to have `alpha == 0` if our canvas is transparent. However, the renderer's `premultipliedAlpha: true` setting means no matter what we use for `setClearColor`, we always get black, since it's multiplied by zero. 

I just need to get the background cleared to (1, 1, 1, 0) - might it be acceptable to add a `premultipliedAlpha` override input to `setClearColor`?

File: src/renderers/shaders/ShaderChunk/transmission_pars_fragment.glsl.js [hunk 2]
```diff
@@ -184,12 +184,15 @@ export default /* glsl */`
 		// Sample framebuffer to get pixel the refracted ray hits.
 		vec4 transmittedLight = getTransmissionSample( refractionCoords, roughness, ior );
 
-		vec3 attenuatedColor = applyVolumeAttenuation( transmittedLight.rgb, length( transmissionRay ), attenuationColor, attenuationDistance );
+		vec3 transmittance = diffuseColor * volumeAttenuation( length( transmissionRay ), attenuationColor, attenuationDistance );
+		vec3 attenuatedColor = transmittance * mix( vec3( 0.5 ), transmittedLight.rgb, transmittedLight.a );
```
Comment (elalish) [inline src/renderers/shaders/ShaderChunk/transmission_pars_fragment.glsl.js:188]: So here I'm mixing in a neutral amount of light (0.5) to represent the unknown light that will come in from the background color.

File: src/renderers/shaders/ShaderChunk/output_fragment.glsl.js [hunk 3]
```diff
@@ -5,8 +5,8 @@ diffuseColor.a = 1.0;
 
 // https://github.com/mrdoob/three.js/pull/22425
 #ifdef USE_TRANSMISSION
-diffuseColor.a *= material.transmissionAlpha + 0.1;
+diffuseColor.a *= material.transmissionAlpha;
 #endif
 
-gl_FragColor = vec4( outgoingLight, diffuseColor.a );
+gl_FragColor = vec4( outgoingLight * diffuseColor.a, diffuseColor.a );
```
Comment (elalish) [inline src/renderers/shaders/ShaderChunk/output_fragment.glsl.js:11]: In this update I no longer need any changes to the default renderer 🎉, instead applying our `premultipliedAlpha` here. The results look better too - the colors stay more saturated. The only issue I notice is a dark outline due to interpolating toward transparent black: 
image
Not too bad in the grand scheme of things. However, this also made some regression render diffs, so we'll need to look at that.

File: src/renderers/shaders/ShaderChunk/output_fragment.glsl.js [hunk 4]
```diff
@@ -5,8 +5,8 @@ diffuseColor.a = 1.0;
 
 // https://github.com/mrdoob/three.js/pull/22425
```
Comment (LeviPesin) [inline src/renderers/shaders/ShaderChunk/output_fragment.glsl.js:6]: Maybe we should remove this comment now?

File: src/renderers/shaders/ShaderChunk/output_fragment.glsl.js (hunk 3 above)
Comment (WestLangley) [reply src/renderers/shaders/ShaderChunk/output_fragment.glsl.js:11]: `output_fragment` is used by almost every built-in material. You can't just multiply RGB by alpha. You will break every example using transparency.

This is why we have the property `material.premultipliedAlpha`.

Comment (elalish) [reply src/renderers/shaders/ShaderChunk/output_fragment.glsl.js]: Thanks, I switched to enabling `premultipliedAlpha` instead. This got me back to the first draft renders, so if I'd tried this first I probably would have been totally happy with the result. But now I have that 2nd draft render to compare to with more saturation that I like better. I can't quite figure out what's causing the difference.

File: src/renderers/WebGLRenderer.js [hunk 5]
```diff
@@ -1269,6 +1273,11 @@ class WebGLRenderer {
 
 			const currentRenderTarget = _this.getRenderTarget();
 			_this.setRenderTarget( _transmissionRenderTarget );
+
+			_this.getClearColor( _currentClearColor );
+			_currentClearAlpha = _this.getClearAlpha();
+			if ( _currentClearAlpha < 1 ) _this.setClearColor( 0xffffff, 0.8 );
```
Comment (elalish) [inline src/renderers/WebGLRenderer.js:1279]: @WestLangley the only thing I changed about your PR when I pulled it in was to set this clear color from gray to white. I was also able to remove the strange mix I'd added to the fragment shader, so I definitely prefer your approach. It's also yielding better renders.

**Comment by elalish** [comment]
@WestLangley pulling in your changes we can now check your live example with my updates: https://raw.githack.com/elalish/three.js/transparency/examples/webgl_materials_physical_transmission.2.html

Note that `material.transparent` is only turned on now when opacity < 1, as this makes a fairly significant rendering difference on its own.

**Comment by elalish** [comment]
I like the 80% minimum opacity for transmissive objects (pulled from #25881) - it's pretty much the only arbitrary constant in here now and it makes a pretty nice compromise between retaining object saturation and still showing the background:
image

**Comment by mrdoob** [comment]
Adding screenshots for my own reference in the future.

| This PR | WL's PR |
| --- | --- |
| Screenshot 2023-04-27 at 11 57 06 | Screenshot 2023-04-27 at 11 57 12 |
| Screenshot 2023-04-27 at 11 58 23 | Screenshot 2023-04-27 at 11 58 20 |
| Screenshot 2023-04-27 at 11 57 28 | Screenshot 2023-04-27 at 11 57 34 |

**Comment by WestLangley** [comment]
Here is an example with a CSS background (not a skybox).

WestLangley's PR

westlangley

This PR

elalish

**Comment by elalish** [comment]
@WestLangley Further improvements would be great! Would you mind responding to the main feature I added compared to yours? https://github.com/mrdoob/three.js/pull/25881#issuecomment-1520638999

File: src/renderers/WebGLRenderer.js (hunk 5 above)
Comment (WestLangley) [reply src/renderers/WebGLRenderer.js:1279]: I do think my PR (updated in #26003), which uses the original gray color `0x7f7f7f`, is preferable. In my testing of a variety of models with only a CSS background, white is too bright.

File: src/renderers/shaders/ShaderChunk/transmission_pars_fragment.glsl.js (hunk 1 above)
Comment (WestLangley) [reply src/renderers/shaders/ShaderChunk/transmission_pars_fragment.glsl.js:195]: I am not sure this is needed.

It is best not to hack the physical model if possible.

File: src/renderers/webgl/WebGLPrograms.js [hunk 6]
```diff
@@ -307,7 +307,7 @@ function WebGLPrograms( renderer, cubemaps, cubeuvmaps, extensions, capabilities
 			toneMapping: material.toneMapped ? renderer.toneMapping : NoToneMapping,
 			useLegacyLights: renderer.useLegacyLights,
 
-			premultipliedAlpha: material.premultipliedAlpha,
+			premultipliedAlpha: material.premultipliedAlpha || HAS_TRANSMISSION,
```
Comment (WestLangley) [inline src/renderers/webgl/WebGLPrograms.js:310]: I am not sure this is justifiable...

Comment (elalish) [reply src/renderers/webgl/WebGLPrograms.js:310]: Good catch, I think I forgot to revert this.

File: src/renderers/shaders/ShaderChunk/transmission_pars_fragment.glsl.js (hunk 1 above)
Comment (elalish) [reply src/renderers/shaders/ShaderChunk/transmission_pars_fragment.glsl.js:195]: This is addressing a specific opacity bug with your alternative approach: https://github.com/mrdoob/three.js/pull/25881#issuecomment-1520638999

In particular, without this, 20% of the background shows through even if the transmissive material is absorbing all or nearly all of the transmitted light. This hack still has no effect when the final rendered output is opaque.

File: src/renderers/WebGLRenderer.js (hunk 5 above)
Comment (elalish) [reply src/renderers/WebGLRenderer.js:1279]: You're probably right - let me check a few more examples.

**Comment by elalish** [comment]
@WestLangley Good point, but this is an unavoidable tradeoff of our blend function restrictions. I'm going to focus only on our difference in clear color, since my other adjustment to allow dark transparent objects to approach opaqueness is orthogonal and only further improves the results (would love to discuss on a separate thread if you're unconvinced). 

I'm proposing a compromise: `setClearColor( 0xffffff, 0.5 )`, which I think is a decent middle ground between working for light backgrounds and working for dark backgrounds. For dark backgrounds, reducing either the color or the alpha moves the result in a good direction. For light backgrounds, the opposite is true, but reducing the color has a more extreme negative impact than reducing the alpha. I also like this compromise because it only involves 1 arbitrary constant rather than two (since multiplying by `#ffffff` is a no-op). 

`setClearColor( 0xffffff, 0.5 )` (this PR):
imageimage
`setClearColor( 0x7f7f7f, 0.8 )` (#26003):
imageimage
Regarding your example (or as close as I could manage): agreed that my compromise is doesn't do as well for it, but it's at least improved a bit and I still think it is outweighed by the examples above.
image

**Comment by donmccurdy** [comment]
I don't care much about the default setting here, but I do care about parameterization, and I haven't been able to follow that from these screenshots and discussions.

As a user, if I have a CSS background that is `#E96143`, can I assume that `setClearColor( 0xE96143, alpha )`, for some fixed alpha will get me a good result under one or both of these PRs? Or is trial and error required to arrive at good choices of one or both parameters?

**Comment by elalish** [comment]
@donmccurdy This `setClear` is actually buried in the renderer and not available in the public API. The idea is that this compromise value does a decent job for *all* CSS backgrounds (though obviously not perfect). If you know your background ahead of time, you're probably better off just applying it as a real three.js background so the rendering is really correct. This PR allows it to still work pretty well when the 3D developer and the CSS developer don't know about each other.

**Comment by donmccurdy** [comment]
@elalish oops sorry! I noticed my mistake and edited my comment while you were responding.

> If you know your background ahead of time...

I'm really thinking of cases where the background is *not* a solid color, but where you can still squint and say it's broadly such and such a color. Certainly if my background is black-and-white checkers, I'm out of luck.

**Comment by elalish** [comment]
@donmccurdy That's an interesting idea. I was trying to make this PR minimally-invasive, so no one would have to change their code or add anything to get the benefit. Having configuration might be nice, but probably something to discuss in a follow-on PR? I actually found that picking the clear alpha to match the overall lightness of the color gave decent results too.

Any thoughts @mrdoob?

**Comment by donmccurdy** [comment]
Sure, I don't mean to block either PR! It looked like we were stuck in an argument about whether to choose defaults that look good on black backgrounds or on white backgrounds, so I thought perhaps passing through the user's clear color to the transmissive pass might unblock things... 😇

**Comment by WestLangley** [comment]
>Regarding your example (or as close as I could manage): agreed that my compromise is doesn't do as well for it,

I agree, it doesn't.

The use case my PR was developed for is one in which the entire background is transparent, and the camera background or CSS shows through.

File: src/renderers/WebGLRenderer.js [hunk 7]
```diff
@@ -1269,6 +1273,11 @@ class WebGLRenderer {
 
 			const currentRenderTarget = _this.getRenderTarget();
 			_this.setRenderTarget( _transmissionRenderTarget );
+
+			_this.getClearColor( _currentClearColor );
+			_currentClearAlpha = _this.getClearAlpha();
+			if ( _currentClearAlpha < 1 ) _this.setClearColor( 0xffffff, 0.5 );
+
```
Comment (WestLangley) [inline src/renderers/WebGLRenderer.js:1280]: This is a change from my PR. Are you aware that the render target will have premultiplied values?

So the two values of interest in this PR are `0xffffff * 0.5 = 0x7f7f7f` and alpha = `0.5`.

In my PR, the values are `0x7f7f7f * 0.8 = 0x666666` and alpha = `0.8`.

Comment (elalish) [reply src/renderers/WebGLRenderer.js:1280]: Good call, we talked offline. I think we're at consensus now. Correct me if I'm wrong.

Comment (WestLangley) [reply src/renderers/WebGLRenderer.js:1280]: This is a subjective decision, and can be changed later.

**Review by WestLangley — APPROVED** [review]

**Comment by elalish** [comment]
> Do you mind removing the builds from the PR? And reverting the changes in `webgl_loader_gltf_transmission`?

Oh, good call; done!

**Comment by mrdoob** [comment]
@elalish FYI

https://github.com/mrdoob/three.js/commit/79cc93ceb85530fc3b16f4e4da6b38d2f658554a#diff-6f5dde4e2eee5725efad5ce991ebe25cbf3c000fa3e26f84a11292fe788f661fL147-L152

**Pull request merged**